keps/sig-auth/4317-pod-certificates/README.md

```diff
@@ -0,0 +1,1440 @@
+
+#### API Server Pod Client Certificate Approver / Signer
+
+The kube-controller-manager is extended to support a new certificate signer,
+`kubernetes.io/kube-apiserver-client-pod`.
+
+The controller for this signer name automatically issues certificates for
+PodCertificateRequests that are addressed to it.  It performs no additional
+validation on PodCertificateRequests beyond the validation enforced by
+kube-apiserver.
+
+The certificates it issues are meant to be understood by kube-apiserver as
+described above.  They:
+* Have a CommonName set to strong random text.  (This maintains flexibility, in
+  case we wish to include a meaningful common name in the future.)
+* Have a PodIdentity extension (non-critical) that contains the information from
+  `PodCertificateRequest.Spec`.
+* Key usages consistent with the certificate being used as a TLS client
+  certificate or server certificate.
+* It uses a CA specified by a new pair of private key / root certificate flags
+  (although implementations are free to set this flag to point at the CA as
+  another signer).
+* It issues certificates that are valid for 24 hours, and tells Kubelet to begin
+  refreshing them at 12 hours.
+
+During Alpha, this feature is gated by the
+`PodCertificateKubeApiserverClientIssuance` feature gate.
+
+Because kube-apiserver will interpret the PodIdentity extension as expressing an
+object binding between the presented certificate and the pod named in the
+extension, once a pod is no longer running, all certificates that were issued to
+it will become invalid.  This, along with the short lifetime of the
+certificates, removes the need for a more permanent certificate revocation
+mechanism.
+
+#### Client-go InClusterConfig Enhancements
+
+If client-go `InClusterConfig()` detects a credential bundle file at
+`/var/run/secrets/kubernetes.io/serviceaccount/credentialbundle.pem`, it will
+configure itself to use the private key and client certificate chain from the
+bundle, periodically reloading them to catch when they are rotated.
+
+### Risks and Mitigations
+
+**Scalability**: In the current design and draft implementation, kubelet holds
+the private keys for each pod on the node in memory.  If kubelet restarts, then
+the private keys are lost, and kubelet will need to immediately generate new
+keys and request new certificates for each volume on the pod.  If kubelet
+crashloops, this can generate high CertificateSigningRequest traffic on
+kube-apiserver.  This can potentially be mitigated by giving kubelet the
+capability to read back the existing key from the projected volume on startup.
+
+### User Stories
+
+#### Story 1
+
+I'm an application developer building a Kubernetes controller that I want to
+deploy into my cluster.  I want to make sure that my controller securely
+authenticates to kube-apiserver using mTLS.
+
+I add a new projected volume to my pod, containing the information that
+client-go expects at a well-known path.  I set
```

Comment:
It seems really awkward to require this very manual process, especially with magic well known paths on disk in play.

Reply:
In the future, we can think about automatically injecting these as a replacement for the existing kube-apiserver-access volumes.  I didn't want to cover that in Alpha because there's already enough work just landing the machinery without thinking about driving adoption.

However, I think there's a larger conversation to be had on this topic in SIG Auth.  My personal thoughts are:
* Automatically granting API access credentials by default is probably a mistake.  Most workloads in a Kubernetes cluster do not actually need to talk to the Kubernetes API.
* Mutating admission controllers (especially webhooks) cause a lot of problems.  In general, workload authors expect the spec they write to be the spec that gets loaded into the cluster.
* In a perfect world, I would make workload authors who want Kubernetes API access opt in by inserting a token (or pod certificate volume).

Reply:
Yeah I am fine with this as-is, but lets discuss further in a future meeting.